server/src/main/java/com/cloud/storage/download/DownloadMonitorImpl.java

```diff
@@ -152,28 +158,34 @@ private void initiateTemplateDownload(DataObject template, AsyncCompletionCallba
-                s_logger.warn(errMsg);
-                throw new CloudRuntimeException(errMsg);
-            }
-            DownloadListener dl = new DownloadListener(ep, store, template, _timer, this, dcmd, callback);
-            ComponentContext.inject(dl);  // initialize those auto-wired field in download listener.
-            if (downloadJobExists) {
-                // due to handling existing download job issues, we still keep
-                // downloadState in template_store_ref to avoid big change in
-                // DownloadListener to use
-                // new ObjectInDataStore.State transition. TODO: fix this later
-                // to be able to remove downloadState from template_store_ref.
-                s_logger.info("found existing download job");
-                dl.setCurrState(vmTemplateStore.getDownloadState());
-            }
+                if (tmpl.getTemplateType() == Storage.TemplateType.SYSTEM) {
+                    managementServerDownloader.download(tmpl, template);
+                    //set to ready state
+                } else {
+                    String errMsg = "There is no secondary storage VM for downloading template to image store " + store.getName();
+                    s_logger.warn(errMsg);
+                    throw new CloudRuntimeException(errMsg);
+                }
+            } else {
+                DownloadListener dl = new DownloadListener(ep, store, template, _timer, this, dcmd, callback);
+                ComponentContext.inject(dl);  // initialize those auto-wired field in download listener.
+                if (downloadJobExists) {
+                    // due to handling existing download job issues, we still keep
+                    // downloadState in template_store_ref to avoid big change in
+                    // DownloadListener to use
+                    // new ObjectInDataStore.State transition. TODO: fix this later
+                    // to be able to remove downloadState from template_store_ref.
+                    s_logger.info("found existing download job");
+                    dl.setCurrState(vmTemplateStore.getDownloadState());
+                }
 
-            try {
-                ep.sendMessageAsync(dcmd, new UploadListener.Callback(ep.getId(), dl));
-            } catch (Exception e) {
-                s_logger.warn("Unable to start /resume download of template " + template.getId() + " to " + store.getName(), e);
-                dl.setDisconnected();
-                dl.scheduleStatusCheck(RequestType.GET_OR_RESTART);
+                try {
+                    ep.sendMessageAsync(dcmd, new UploadListener.Callback(ep.getId(), dl));
+                } catch (Exception e) {
+                    s_logger.warn("Unable to start /resume download of template " + template.getId() + " to " + store.getName(), e);
+                    dl.setDisconnected();
+                    dl.scheduleStatusCheck(RequestType.GET_OR_RESTART);
+                }
```

Comment:
the structure here has changed: no longer will ep be checked for null and fixed, but now ep will be either bypassed or used. This means that handling of the download process is different for the new case. and thus code duplication is introduced somewhere.
Can you comment?